modules/io/adaptors/parse_bytes_to_dataframe.cc

```diff
@@ -51,31 +55,80 @@ void ParseTable(std::shared_ptr<arrow::Table>* table,
 
   auto read_options = arrow::csv::ReadOptions::Defaults();
   auto parse_options = arrow::csv::ParseOptions::Defaults();
+
+  read_options.column_names = original_columns;
+  parse_options.delimiter = delimiter;
+
   auto convert_options = arrow::csv::ConvertOptions::Defaults();
 
-  if (col_names.size() > 0) {
-    read_options.column_names = col_names;
-  } else {
-    read_options.autogenerate_column_names = (!header_row);
+  auto is_number = [](const std::string& s) -> bool {
+    return !s.empty() && std::find_if(s.begin(), s.end(), [](unsigned char c) {
+                           return !std::isdigit(c);
+                         }) == s.end();
+  };
+
+  std::vector<int> indices;
+  for (size_t i = 0; i < columns.size(); ++i) {
+    if (is_number(columns[i])) {
+      int col_idx = std::stoi(columns[i]);
+      if (col_idx >= static_cast<int>(original_columns.size())) {
+        return Status(StatusCode::kArrowError,
+                      "Index out of range: " + columns[i]);
+      }
+      indices.push_back(col_idx);
+      columns[i] = original_columns[col_idx];
+    }
   }
-  parse_options.delimiter = delimiter;
+
+  // If include_all_columns_ is set, push other names as well
+  if (include_all_columns) {
+    for (const auto& col : original_columns) {
+      if (std::find(std::begin(columns), std::end(columns), col) ==
+          columns.end()) {
+        columns.push_back(col);
+      }
+    }
+  }
+
+  convert_options.include_columns = columns;
+
+  if (column_types.size() > convert_options.include_columns.size()) {
+    return Status(StatusCode::kArrowError,
+                  "Format of column type schema is incorrect.");
+  }
+  std::unordered_map<std::string, std::shared_ptr<arrow::DataType>>
+      arrow_column_types;
+
+  for (size_t i = 0; i < column_types.size(); ++i) {
+    if (!column_types[i].empty()) {
+      arrow_column_types[convert_options.include_columns[i]] =
+          type_name_to_arrow_type(column_types[i]);
+    }
+  }
+  convert_options.column_types = arrow_column_types;
 
   std::shared_ptr<arrow::csv::TableReader> reader;
-  CHECK_ARROW_ERROR_AND_ASSIGN(
+  RETURN_ON_ARROW_ERROR_AND_ASSIGN(
       reader, arrow::csv::TableReader::Make(pool, input, read_options,
                                             parse_options, convert_options));
 
-  CHECK_ARROW_ERROR_AND_ASSIGN(*table, reader->Read());
+  auto result = reader->Read();
+  if (!result.status().ok()) {
+    if (result.status().message() == "Empty CSV file") {
+      *table = nullptr;
+      return Status::OK();
+    } else {
+      return Status::ArrowError(result.status());
+    }
+  }
+  *table = result.ValueOrDie();
```

Comment:
Use `RETURN_ON_ARROW_ERROR_AND_ASSIGN`,  `ValueOrDie` will `abort` and we have no change for error processing.